### ISSUE TITLE
Suggest NodeTreeSanitizer.trusted if using constant text with innerHtml= or appendHtml

### ISSUE THREAD

**Issue by alan-knight** [dart-lang/linter#109]
In 1.12 we introduce a more efficient way for sanitizing trusted HTML in the dart:html library. There are a lot of uses of the APIs that can take a sanitizer which pass constant HTML, which presumably can be trusted. So a lint suggestion do

``` dart
  someElement.appendHtml('<p>Hello</p>', treeSanitizer: NodeTreeSanitizer.trusted);
```

rather than

```
  someElement.appendHtml('<p>Hello</p>');
```

or even rather than using their own NullSanitizer class, might improve performance significantly.

The APIs that take this are
- the html constructor
- appendHtml()
- setInnerHtml
-  insertAdjacentHtml

These are implemented on 
- Element
- DocumentFragment

**Comment by pq** [comment]
Is there *ever* a time when the default sanitizer is OK/preferred?  What about validators?

**Comment by alan-knight** [comment]
If you didn't write/don't trust the HTML, then you'd want the default sanitizer, or a more complex custom one. The consequences of appending untrusted HTML without sanitizing it are bad.

**Comment by pq** [comment]
Got it.  My concern with a lint is false positives.  Any ideas for added heuristics we could use to avoid any?  Has anyone thought about something like a `@trusted` annotation?

**Comment by alan-knight** [comment]
Not HTML sanitization, but a similar concept :-) https://xkcd.com/327/

The heuristic I was thinking of was that if it's a literal (without interpolation) then you wrote it yourself in your source code and it's probably OK. And that's fairly common pattern. The other is if someone wrote their own NullSanitizer (and they always seem to call it NullSanitizer) then switching to this would be more efficient.

**Comment by pq** [comment]
@alan-knight: I'm thinking we no longer want to pursue this... Feel free to re-open if you think otherwise!
